pkg/controller/pxc/users.go

```diff
@@ -0,0 +1,284 @@
+
+	if !dataChanged {
+		return nil
+	}
+
+	var restartPXC bool
+	var restartProxy bool
+
+	err = r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj, &restartPXC, &restartProxy)
+	if err != nil {
+		return errors.Wrap(err, "manage sys users")
+	}
+
+	err = r.updateInternalSysUsersSecret(cr, &sysUsersSecretObj)
+	if err != nil {
+		return errors.Wrap(err, "update internal sys users secret")
+	}
+
+	if restartProxy && cr.Spec.ProxySQL != nil && cr.Spec.ProxySQL.Size > 0 {
+		err = r.restartProxy(cr, newSecretDataHash)
+		if err != nil {
+			return errors.Wrap(err, "restart proxy")
+		}
+	}
+
+	if restartPXC {
+		err = r.restartPXC(cr, newSecretDataHash, &sysUsersSecretObj)
+		if err != nil {
+			return errors.Wrap(err, "restart pxc")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) manageSysUsers(cr *api.PerconaXtraDBCluster, sysUsersSecretObj, internalSysSecretObj *corev1.Secret, restartPXC, restartProxy *bool) error {
+	um, err := users.NewManager([]string{cr.Name + "-pxc"}, string(internalSysSecretObj.Data["root"]))
+	if err != nil {
+		return errors.Wrap(err, "new users manager")
+	}
+	var sysUsers []users.SysUser
+
+	for name, pass := range sysUsersSecretObj.Data {
+		hosts := []string{}
+
+		if string(sysUsersSecretObj.Data[name]) == string(internalSysSecretObj.Data[name]) {
+			continue
+		}
+
+		switch name {
+		case "root":
+			*restartProxy = true
+			hosts = []string{"localhost", "%"}
+		case "xtrabackup":
+			*restartProxy = true
+			*restartPXC = true
+			hosts = []string{"localhost"}
+		case "monitor":
+			*restartProxy = true
+			hosts = []string{"10.%", "%"}
+		case "clustercheck":
+			*restartProxy = true
+			*restartPXC = true
+			hosts = []string{"localhost"}
+		case "proxyadmin":
+			*restartProxy = true
+			continue
+
+		case "pmmserver":
+			if cr.Spec.PMM.Enabled {
+				*restartProxy = true
+				*restartPXC = true
+				continue
+			}
+		}
+		user := users.SysUser{
+			Name:  name,
+			Pass:  string(pass),
+			Hosts: hosts,
+		}
+		sysUsers = append(sysUsers, user)
+	}
+
+	if len(sysUsers) > 0 {
+		err = um.UpdateUsersPass(sysUsers)
+		if err != nil {
+			return errors.Wrap(err, "update sys users pass")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) restartPXC(cr *api.PerconaXtraDBCluster, newSecretDataHash string, sysUsersSecretObj *corev1.Secret) error {
+	sfsPXC := appsv1.StatefulSet{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Namespace: cr.Namespace,
+			Name:      cr.Name + "-pxc",
+		},
+		&sfsPXC,
+	)
+
+	if len(sfsPXC.Annotations) == 0 {
+		sfsPXC.Annotations = make(map[string]string)
+	}
+	sfsPXC.Spec.Template.Annotations["last-applied-secret"] = newSecretDataHash
+
+	err = r.client.Update(context.TODO(), &sfsPXC)
+	if err != nil {
+		return errors.Wrap(err, "update pxc sfs last-applied annotation")
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) restartProxy(cr *api.PerconaXtraDBCluster, newSecretDataHash string) error {
```

Comment:
`newSecretDataHash` is unused

Reply:
removed https://github.com/percona/percona-xtradb-cluster-operator/pull/418/commits/9a770955143b4e0e1b00e53653fdb45ab333357e